src/HRayData.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE AllowAmbiguousTypes       #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE RecordWildCards           #-}
+{-# LANGUAGE ScopedTypeVariables       #-}
+
+module HRayData
+    ( AABB
+    , AnyHit
+    , BoundingVolume
+    , ClosestHit
+    , Color
+    , Context
+    , D2
+    , D3
+    , Intersect
+    , Intersection
+    , Ray
+    , RayResult
+    , Scene
+    , SceneObject
+    -- | Functions
+    , getDouble
+    , getVector2
+    , getVector3
+    , raycast
+    ) where
+
+import Data.Map (Map, (!))
+import Linear.V2 (V2)
+import Linear.V3 (V3)
+
+import qualified Data.Map as Map hiding (Map)
+import qualified Linear.V2 as V2 hiding (V2)
+import qualified Linear.V3 as V3 hiding (V3)
+
+newtype D2 =
+    D2 (V2 Double)
+
+newtype D3 =
+    D3 (V3 Double)
+
+newtype Color =
+    Color D3
+
+data Ray = Ray
+    { origin    :: D3
+    , direction :: D3
+    }
+
+newtype Intersection =
+    Intersection D3
+
+-- | Ray Result is either a Color
+-- or a ray generator and a function to convert the ray results into Color
+data RayResult
+    = RayResult Color
+    | RayResultAssembly [Ray]
+                        ([RayResult] -> Color)
+
+newtype Intersect =
+    Intersect (Context -> Maybe Intersection)
+
+-- | Called when the intersection point was the closest.
+newtype ClosestHit =
+    ClosestHit RayResult
+
+-- | Called when looking for any intersection point.
+-- Returns a RayResult to stop loooking for intersection points or Nothing to keep looking.
+newtype AnyHit =
+    AnyHit (Maybe RayResult)
+
+data Context = Context
+    { v1s :: Map String Double
+    , v2s :: Map String D2
+    , v3s :: Map String D3
+    }
+
+-- | Is used to determine if an underlying SceneObject has a chance of intersecting with the ray.
+class BoundingVolume b where
+    intersect :: Intersect
+
+-- | Axis Alligned Bounding Box.
+data AABB = AABB
+    { min :: D3
+    , max :: D3
```

Comment:
As you know, Haskell has problems with record names. So for now it's better to add prefixes of types to fields to avoid collisions. Also, `min` and `max` are bad names since there're such functions in `Prelude`.